addons/test_mail_full/tests/test_portal.py

```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
 
         record_portal._portal_ensure_token()
         self.assertTrue(record_portal.access_token)
+
+
+@tagged('portal')
+class TestPortalFlow(MailCommon, HttpCase):
+    """Share a link by email to a customer without an account for viewing a record through the portal.
+
+    The tests consist in sending a mail related to a record to a customer and checking that the record can be viewed
+    through the embedded link:
+    - either in the backend if the user is connected and has the right to
+    - or in the portal otherwise
+    """
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.customer = cls.env['res.partner'].create({
+            'name': 'Wood Corner',
+            'email': 'wood_corner@example.com',
+        })
+        cls.record_customer = cls.env['mail.test.portal'].create({
+            'partner_id': cls.customer.id,
+            'name': 'Test Portal Record',
+            'user_id': cls.user_admin.id,
+        })
+        cls.mail_template = cls.env['mail.template'].create({
+            'name': 'Test send quotation',
+            'subject': 'Your quotation',
+            'body_html': '<p>Hello, your quotation is ready for review.</p>',
+            'auto_delete': True,
+            'email_from': '{{ (object.user_id.email_formatted or user.email_formatted) }}',
+            'partner_to': '{{ object.partner_id.id }}',
+            'model_id': cls.env.ref('test_mail_full.model_mail_test_portal').id,
+        })
+        cls._create_portal_user()
+        # Get access URL that is embedded in the customer email for accessing the record.
+        for group_name, group_func, group_data in cls.record_customer.sudo()._notify_get_recipients_groups():
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        # Transform that access URL to have one with an invalid access token
+        cls.record_access_url_wrong_token = re.sub(r'access_token=[0-9a-z-]*',
+                                                   'access_token=00000000-0000-0000-0000-000000000000',
+                                                   cls.record_access_url)
+
+    def test_initial_data(self):
+        """Test some initial value."""
+        self.assertNotEqual(self.record_access_url, self.record_access_url_wrong_token)
```

Comment:
Could be great to parse the url and check the URL itself (containing /my/test_portal/rec.id + token (and maybe pid and hash, depending on how we call things, don't remember all details of the call stack ^^).

Reply:
Yes, indeed, that’s better to do a more in-depth validation of init data.